internal/dcache/rpc/server/handler.go

```diff
@@ -1878,6 +1879,241 @@ refreshFromClustermapAndRetry:
 	return resp, nil
 }
 
+// PutChunkEx RPC takes the PutChunkRequest and a list of next RVs to forward the request to as parameters.
+// The PutChunkRequest is for the local RV. If the RV mentioned in the PutChunkRequest is not the local RV,
+// it will return an InvalidRVID error.
+// Parallelly, it also forwards the PutChunkRequest to the next RVs in the list.
+//
+// For the local RV, it calls the PutChunk RPC via the handler directly.
+// The errors returned by the PutChunk calls to the local RVs and next RVs are returned in the
+// PutChunkExResponse, with the RV name as the key and its PutChunkResponse or error as the value.
+// The PutChunkExResponse will have the responses for all the RVs in the list, including the local RV.
+// If the PutChunkExRequest is for an RV that is not hosted by this node, it will return an InvalidRVID error.
+func (h *ChunkServiceHandler) PutChunkEx(ctx context.Context, req *models.PutChunkExRequest) (*models.PutChunkExResponse, error) {
+	// Thrift should not be calling us with nil req.
+	common.Assert(req != nil)
+
+	// Thrift should not be calling us with nil Request, Chunk or Address.
+	common.Assert(req.Request != nil)
+	common.Assert(req.Request.Chunk != nil)
+	common.Assert(req.Request.Chunk.Address != nil)
+
+	log.Debug("ChunkServiceHandler::PutChunkEx: Received PutChunkEx request: %v",
+		rpc.PutChunkExRequestToString(req))
+
+	// rvID must refer to one of of out local RVs.
+	rvInfo, ok := h.rvIDMap[req.Request.Chunk.Address.RvID]
+	if !ok || rvInfo == nil {
+		errStr := fmt.Sprintf("Invalid rvID %s", req.Request.Chunk.Address.RvID)
+		log.Err("ChunkServiceHandler::PutChunkEx: %s", errStr)
+		common.Assert(false, errStr)
+		return nil, rpc.NewResponseError(models.ErrorCode_InvalidRVID, errStr)
+	}
+
+	var rpcResp *models.PutChunkExResponse
+	var err error
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+
+	//
+	// Parallelly forward the PutChunkEx request to the next RVs in the list.
+	//
+	go func() {
+		defer wg.Done()
+
+		rpcResp = h.forwardPutChunk(ctx, req.Request, req.NextRVs)
```

Comment:
we should not need forwardPutChunk(), instead PutChunkEx should be implemented recursively as

PutChunk for the Request entry
PutChunkEx with nextRVs[0] copied to the Request and nextRVs[1:] as the new nextRV. If there's only one element left in nextRV, then just make a PutChunk call for that.

Response should be gathered and sent for all the RVs (one in Request and those in nextRVs)